src/timer.c

```diff
@@ -15,141 +13,80 @@
-					sm_cPrecision = CLOCKS_PER_SEC / 1000;
-			}
-			else {
-				if (!sm_cPrecision || sm_cPrecision > CLOCKS_PER_SEC / 10000)
-					sm_cPrecision = CLOCKS_PER_SEC / 10000;
-			}
-		}
-
-		// Find the claimed resolution of the high res timer
-		// Then find the most likely real rate by waiting for it to change.
-		// Note - I've frequently seen missed beats, and therefore a
-		// 0.000001 reality gets reported as a 0.000002.
-		// Note - this also isn't actually used, all that matters is
-		// whether HRTicksPerSec has a non-zero value or not.
-		HRGETTICKS_PER_SEC(sm_HRTicksPerSec);
-
-		if (sm_HRTicksPerSec) {
-			hr_timer start, end;
-
-			HRSETCURRENT(start);
-			do {
-				HRSETCURRENT(end);
-			} while (HRGETTICKS(end) == HRGETTICKS(start));
-
-			sm_hrPrecision = sm_HRTicksPerSec / (HRGETTICKS(end) -
-			                  HRGETTICKS(start));
-		}
-	}
-}
+	int num = 0;
+	uint64_t start, end;
 
+	if (!timer_ticks_per_sec)
+		HRSETRESOLUTION();
 
-void sTimer_Stop(sTimer *t)
-{
-	if (t->m_fRunning) {
-		if (sm_HRTicksPerSec)
-			HRSETCURRENT(t->m_hrEndTime);
-		else
-			t->m_cEndTime = clock();
-	} else
-		HRZERO(t->m_hrEndTime);
-
-	t->m_fRunning = 0;
+	start = timer_get_nano();
+	do {
+		end = timer_get_nano();
+		num++;
+	} while (end == start);
+
+	if (latency)
+		*latency = (num == 1);
+	return end - start;
 }
 
-void sTimer_Start(sTimer *t, int bClear)
+void timer_init(hr_timer_t *t)
 {
-	if (bClear)
-		sTimer_ClearTime(t);
+	if (t) {
+		t->running = 0;
+		t->accumulated = 0;
+		HRZERO(t->start);
+		HRZERO(t->end);
+	}
+
+	if (!timer_ticks_per_sec)
+		HRSETRESOLUTION();
+}
 
-	if (sm_HRTicksPerSec)
-		HRSETCURRENT(t->m_hrStartTime);
+void timer_stop(hr_timer_t *t)
+{
+	if (t->running)
+		HRSETCURRENT(t->end);
 	else
-		t->m_cStartTime = clock();
+		HRZERO(t->end);
 
-	t->m_fRunning = 1;
+	t->running = 0;
 }
 
-void sTimer_ClearTime(sTimer *t)
+void timer_start(hr_timer_t *t, int clear)
 {
-	t->m_dAccumSeconds = 0;
-	HRZERO(t->m_hrStartTime);
-	HRZERO(t->m_hrEndTime);
+	if (clear)
+		timer_clear(t);
 
-	t->m_fRunning = 0;
+	timer_start_noclear(t);
 }
 
-double sTimer_GetSecs(sTimer *t)
+void timer_clear(hr_timer_t *t)
+{
+	t->accumulated = 0;
+	HRZERO(t->start);
+	HRZERO(t->end);
+
+	t->running = 0;
+}
+
+double timer_get_secs(hr_timer_t *t)
 {
 	double retval;
 
-	if (t->m_fRunning) {
-		if (sm_HRTicksPerSec)
-			HRSETCURRENT(t->m_hrEndTime);
-		else
-			t->m_cEndTime = clock();
-	}
-	if (!sm_HRTicksPerSec) {
-		// This is process time
-		double d = (t->m_cEndTime - t->m_cStartTime);
-
-		if (d > 0)
-			retval = d / CLOCKS_PER_SEC;
-		else
-			retval = 0;
-	}
-	else {
-		// This is wall-clock time
-		double d = (HRGETTICKS(t->m_hrEndTime) - HRGETTICKS(t->m_hrStartTime));
-
-		if (d > 0)
-			retval = d / sm_HRTicksPerSec;
-		else
-			retval = 0;
-	}
+	if (t->running)
+		HRSETCURRENT(t->end);
+
+	double d = HRGETTICKS(t->end) - HRGETTICKS(t->start);
+
+	retval = d / timer_ticks_per_sec;
```

Comment:
BTW, we just accept the risk of division by zero (in multiple places) if certain system-provided timers return weird resolution, right? No way to handle this better?

Reply:
Sure but not without the cost of latency, which I wanted to keep down to a minimum. I "ruled out" the case of system returning weird resolution because I believe it can't happen unless you run Windows 95 or something that ridiculous and I really doubt you'll manage to even build current Jumbo on that. Another fact is if you use the get_nano macro without calling init first, you'll get a div by zero segfault. Perhaps there should at least be comments describing the requirements. On the other hand, maybe I'm just silly and adding a conditional to set the resolution unless not already set is a very tiny fraction of the latency we have anyway.